cmd/metal-api/internal/fsm/fsm_test.go

```diff
@@ -0,0 +1,315 @@
+				Liveliness:           metal.MachineLivelinessAlive,
+				LastEventTime:        &now,
+				Events: metal.ProvisioningEvents{
+					{
+						Time:  now,
+						Event: metal.ProvisioningEventPXEBooting,
+					},
+				},
+			},
+		},
+		{
+			name: "Transition from PXEBooting to PXEBooting",
+			container: &metal.ProvisioningEventContainer{
+				Events: metal.ProvisioningEvents{
+					{
+						Time:  lastEventTime,
+						Event: metal.ProvisioningEventPXEBooting,
+					},
+				},
+				Liveliness: metal.MachineLivelinessAlive,
+			},
+			event: &metal.ProvisioningEvent{
+				Time:  now,
+				Event: metal.ProvisioningEventPXEBooting,
+			},
+			wantErr: nil,
+			want: &metal.ProvisioningEventContainer{
+				CrashLoop:            false,
+				FailedMachineReclaim: false,
+				Liveliness:           metal.MachineLivelinessAlive,
+				LastEventTime:        &now,
+				Events: metal.ProvisioningEvents{
+					{
+						Time:  lastEventTime,
+						Event: metal.ProvisioningEventPXEBooting,
+					},
+				},
+			},
+		},
+		{
+			name: "Transition from PXEBooting to Preparing",
+			container: &metal.ProvisioningEventContainer{
+				Events: metal.ProvisioningEvents{
+					{
+						Time:  lastEventTime,
+						Event: metal.ProvisioningEventPXEBooting,
+					},
+					{
+						Time:  lastEventTime.Add(time.Minute * 2),
+						Event: metal.ProvisioningEventPXEBooting,
+					},
+				},
+				Liveliness: metal.MachineLivelinessAlive,
+			},
+			event: &metal.ProvisioningEvent{
+				Time:  now,
+				Event: metal.ProvisioningEventPreparing,
+			},
+			wantErr: nil,
+			want: &metal.ProvisioningEventContainer{
+				CrashLoop:            false,
+				FailedMachineReclaim: false,
+				Liveliness:           metal.MachineLivelinessAlive,
+				LastEventTime:        &now,
+				Events: metal.ProvisioningEvents{
+					{
+						Time:  lastEventTime,
+						Event: metal.ProvisioningEventPXEBooting,
+					},
+					{
+						Time:  lastEventTime.Add(time.Minute * 2),
+						Event: metal.ProvisioningEventPXEBooting,
+					},
+					{
+						Time:  now,
+						Event: metal.ProvisioningEventPreparing,
+					},
+				},
+			},
+		},
+		{
+			name: "Transition from Booting New Kernel to Phoned Home",
+			container: &metal.ProvisioningEventContainer{
+				Events: metal.ProvisioningEvents{
+					{
+						Time:  lastEventTime,
+						Event: metal.ProvisioningEventBootingNewKernel,
+					},
+				},
+				Liveliness: metal.MachineLivelinessAlive,
+			},
+			event: &metal.ProvisioningEvent{
+				Time:  now,
+				Event: metal.ProvisioningEventPhonedHome,
+			},
+			wantErr: nil,
+			want: &metal.ProvisioningEventContainer{
+				CrashLoop:            false,
+				FailedMachineReclaim: false,
+				Liveliness:           metal.MachineLivelinessAlive,
+				LastEventTime:        &now,
+				Events: metal.ProvisioningEvents{
+					{
+						Time:  lastEventTime,
+						Event: metal.ProvisioningEventBootingNewKernel,
+					},
+					{
+						Time:  now,
+						Event: metal.ProvisioningEventPhonedHome,
+					},
```

Comment:
I think in the past the events in the event container were sorted anti-chronologically (latest to oldest). Would it make sense to also call your new `Validate` function in the test to ensure this order is correct?

Reply:
This is from a current DB:

![Screenshot from 2022-07-26 08-28-03](https://user-images.githubusercontent.com/15035165/180938551-feede873-5717-4561-8976-f6cd9265304c.png)